src/derived.rs

```diff
@@ -228,6 +235,70 @@ impl<DB: Database> std::fmt::Debug for MemoInputs<DB> {
     }
 }
 
+type LinkedHashSet<T> = LinkedHashMap<T, (), BuildHasherDefault<FxHasher>>;
+
+struct QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    lru_keys: LinkedHashSet<Q::Key>,
+    data: FxHashMap<Q::Key, QueryState<DB, Q>>,
+}
+
+impl<DB, Q> Default for QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    fn default() -> Self {
+        QueryMap {
+            lru_keys: Default::default(),
+            data: Default::default(),
+        }
+    }
+}
+
+impl<DB, Q> QueryMap<DB, Q>
+where
+    Q: QueryFunction<DB>,
+    DB: Database,
+{
+    fn set_lru_capacity(&mut self, new_capacity: usize) {
+        if new_capacity == 0 {
+            self.lru_keys.clear();
+        } else {
+            while self.lru_keys.len() > new_capacity {
+                self.remove_lru();
+            }
+            let additional_cap = new_capacity - self.lru_keys.len();
+            self.lru_keys.reserve(additional_cap);
+        }
+    }
+
+    fn record_use(&mut self, key: &Q::Key, lru_cap: usize) {
+        self.lru_keys.insert(key.clone(), ());
+        if self.lru_keys.len() > lru_cap {
+            self.remove_lru();
+        }
+    }
+
+    fn remove_lru(&mut self) {
+        if let Some((evicted, ())) = self.lru_keys.pop_front() {
+            if let Some(QueryState::Memoized(memo)) = self.data.get_mut(&evicted) {
+                // Similar to GC, evicting a value with an untracked input could
+                // lead to inconsistencies. Note that we can't check
+                // `has_untracked_input` when we add the value to the cache,
+                // because inputs can become untracked after the fact.
```

Comment:
Can you elaborate on what you mean by "after the fact"? By the time we record the value, we *do* know its inputs, but I guess you're saying that's too late? Maybe we should loop in the event that `has_untracked_input` is true (i.e., keep trying to remove things?)

Reply:
I guess you could say the same for the other conditions.